Add scrollbar offset theme constants to Tree

timothyqiu opened the pull request:
This PR adds `scrollbar_offset_{left,top,right,bottom}` theme constants to `Tree`. These constants move scrollbar edges toward the correspoinding border.

---

`Tree` allows setting a background stylebox while `ScrollContainer` don't. Scrollbars in `Tree` use stylebox margins when placing the scrollbars.

The screenshot below is taken from Editor Settings. The scrollbar on the left is from `Tree`, and the scrollbar on the right is from `ScrollContainer`.

![before-after](https://user-images.githubusercontent.com/372476/210476380-ecb852ec-4f0f-4af1-b96f-ef17aab03c55.png)

From top to bottom:

1. 4.0beta10: The Y positions are different. (Because editor theme sets content margin in `Tree`'s panel stylebox.)
2. After #70877: The scrollbar honours background stylebox content right margin, which makes the difference more significant.
3. After this PR: Added offset in editor theme so that Tree scrollbars and ScrollContainer scrollbars look the same.

_Production edit: Closes https://github.com/godotengine/godot/issues/73139_

timothyqiu committed:
Add scrollbar offset theme constants to Tree

YuriSizov commented:
I don't like that we resort to hacks to fix our inconsistencies instead of making things more consistent. Personally, I constantly have to work around the limitation of the scroll container, adding a margin container and making it a clipping rect. So I wouldn't mind having more control in both elements here, which gives us an opportunity to harmonize them instead of ad-hoc fixing the Tree in this manner.

I also don't like that offsets do the opposite of their name and positive numbers are moving the scrollbar outwards instead of inwards. Though that's an easy change.

I would detach the position of scrollbars from the stylebox's margins completely in Tree. We can still use the 4 constants, or we can add a new stylebox that would only contribute margins (and thus has to be StyleBoxEmpty). Same in the scroll container, we need a set of margins for the content and another for the scrollbars. Can be 4 constants each, or 2 margin-only styleboxes.

timothyqiu commented:
Updated:

1. Now the constants are named `scrollbar_margin_*`. They push the scrollbars inward, relative to the control's borders. When negative, which is the default, margins from the background `panel` stylebox are used.
2. Make tree content aware of scrollbar offset.
3. Adds `scrollbar_{h,v}_separation` constants to define the gap between tree content and scrollbars. There were originally no distance. The gap you saw was the result of the offset scrollbars.
4. Don't reserve space at the bottom right corner if only one of the scrollbars is visible.

YuriSizov commented on doc/classes/Tree.xml:
```diff
@@ -543,6 +543,24 @@
+			The right margin of the scrollbars. When negative, uses [theme_item panel] right margin.
+		</theme_item>
+		<theme_item name="scrollbar_margin_top" data_type="constant" type="int" default="-1">
+			The right margin of the vertial scrollbar. When negative, uses [theme_item panel] top margin.
```
Typo here and below, "vertical".

YuriSizov commented on scene/gui/tree.cpp:
In `Tree::scroll_to_item` we have `h_scroll->get_combined_minimum_size().height;`. Should it also be update to the new method?

YuriSizov commented on scene/gui/tree.cpp:
```diff
@@ -3188,6 +3195,32 @@ bool Tree::_scroll(bool p_horizontal, float p_pages) {
+	const Size2 size = get_size();
+	const Ref<StyleBox> background = theme_cache.panel_style;
+
+	const Rect2 content_rect = Rect2(background->get_offset(), size - background->get_minimum_size());
```
This one is weird. I think we want to subtract the right and bottom margin instead of the min size. It probably worked by accident before, but StyleBoxes can override the min size now, and it would no longer represent the right-bottom offset. (Same in the original method where this was copied from and where this is still used too).

YuriSizov replied:
I changed my comment here, took me a moment to unwrap what's going on.

timothyqiu replied:
The parameters here are a position and a size. `background->get_minimum_size()` returns the sum of all margins introduced by the stylebox. So I don't think it assumes the equality margins.

YuriSizov replied:
The code here _assumes_ that `background->get_minimum_size()` returns the sum of all margins, but that's not necessarily true. It's the default value/behavior, but it can be changed. So I'm asking you to change this, and other similar, part to subtract the margins explicitly rather than assume that the minimum size always equals to them.

timothyqiu replied:
Makes sense. I'll update this PR tomorrow.

YuriSizov commented on scene/gui/tree.cpp:
```diff
@@ -3188,6 +3195,32 @@ bool Tree::_scroll(bool p_horizontal, float p_pages) {
+Size2 Tree::_get_content_rect_offset() const {
+	const Size2 size = get_size();
+	const Ref<StyleBox> background = theme_cache.panel_style;
+	const Point2 content_end = background->get_offset() + size - background->get_minimum_size();
```
Same here.

YuriSizov replied:
In fact, here we can simply take the size and subtract the right-bottom offset, as the top-left is cancelled out.

YuriSizov commented on scene/gui/tree.cpp:
```diff
@@ -3854,11 +3887,6 @@ void Tree::update_scrollbars() {
 	const Size2 vmin = v_scroll->get_combined_minimum_size();
 
 	const Rect2 content_rect = Rect2(theme_cache.panel_style->get_offset(), size - theme_cache.panel_style->get_minimum_size());
```
Same here.

YuriSizov reviewed (approved):
Okay, I'm done with the review. I think all the math checks out, so LGTM!

Approving tentatively but the comments above need to be resolved. 

PS. I would like to repeat that it'd be nice to do the same to ScrollContainer, but it seems to be a harder task at the moment. Namely because we have a hack for the animation editor, introduced in https://github.com/godotengine/godot/commit/b659fd6d7442701284cbb8763fb712be36d17ed0.

timothyqiu commented:
Updated.

I replaced `_get_content_rect_offset()` with `_get_content_rect()` so many other places calculating with `bg->get_minimum_size()`  and scrollbar sizes can benefit.

YuriSizov reviewed (approved):
This is a great refactoring of the sizing logic, good job!

The fact that items no longer clip the scrollbar is so satisfying:
![godot windows editor dev x86_64_2023-04-07_13-18-30](https://user-images.githubusercontent.com/11782833/230600871-d767590f-39cf-41e2-83e3-b05e849ad03a.png)

YuriSizov commented:
Thanks!
